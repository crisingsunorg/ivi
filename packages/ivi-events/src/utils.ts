import { FEATURES, FeatureFlags } from "ivi-core";
import { NativeEventSourceFlags } from "./flags";
import { EventHandler } from "./event_handler";

/**
 * Extends Element interface with a special property that is used to assign synthetic event handlers list.
 */
declare global {
  interface Element {
    _ev: Array<EventHandler | null> | EventHandler | null | undefined;
  }
}

/**
 * getEventTarget retrieves target element from an event.
 *
 * There are still some differences in modern browsers, so we need to use this function to retrieve event targets.
 *
 * #quirks
 *
 * @param ev Event.
 * @returns Target Element.
 */
export function getEventTarget(ev: Event): EventTarget {
  let target = ev.target || window;

  /**
   * Some browsers are implementing it according to SVG 1.1 specs:
   *
   * https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7998724/
   *
   * "If event attributes are assigned to referenced elements, then the actual target for the event will be the
   * SVGElementInstance object within the 'instance tree' corresponding to the given referenced element."
   *
   * https://www.w3.org/TR/SVG/struct.html#UseElement
   *
   * SVG 2 redefined the use event handling model:
   *
   * - Removed the SVGElementInstance and SVGElementInstanceList interfaces, and the corresponding attributes on the
   *   SVGUseElement interface.
   * - Changed the ‘use’ element event flow to follow the Shadow DOM spec.
   *
   * https://www.w3.org/TR/SVG2/changes.html#structure
   */
  if ((target as any).correspondingUseElement !== undefined) {
    target = (target as any).correspondingUseElement;
  }

  /**
   * Safari fires events on Text Nodes.
   *
   * http://www.quirksmode.org/js/events_properties.html
   */
  return (target as Node).nodeType === 3 ? (target as Node).parentNode! : target;
}

/**
 * `{ capture: true, passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_PASSIVE_OPTIONS =
  ((FEATURES & FeatureFlags.PassiveEvents) !== 0) ?
    { "capture": true, "passive": true } :
    true;

/**
 * `{ passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_PASSIVE_OPTIONS =
  ((FEATURES & FeatureFlags.PassiveEvents) !== 0) ?
    { "passive": true } :
    false;

/**
 * getNativeEventOptions converts `NativeEventSourceFlags` to event options that can be used as a 3rd parameter
 * for `addEventListener` method.
 *
 * #quirks
 *
 * @param flags See `NativeEventSourceFlags` for details.
 * @returns Option object that can be used as a 3rd parameter for `addEventListener` method.
 */
export function getNativeEventOptions(
  flags: NativeEventSourceFlags,
): boolean | { capture?: boolean, passive?: boolean } {
  if ((flags & NativeEventSourceFlags.Passive) !== 0) {
    if ((flags & NativeEventSourceFlags.Capture) !== 0) {
      return EVENT_CAPTURE_PASSIVE_OPTIONS;
    }
    return EVENT_PASSIVE_OPTIONS;
  }
  if ((flags & NativeEventSourceFlags.Capture) !== 0) {
    return true;
  }
  return false;
}

/**
 * setEventHandlersToDOMNode assigns a event handlers to the DOM node.
 *
 * @param node DOM Node.
 * @param events Event Handlers.
 */
export function setEventHandlersToDOMNode(
  node: Element,
  events: Array<EventHandler | null> | EventHandler | null,
): void {
  node._ev = events;
}

/**
 * getEventHandlersFromDOMNode retrieves event handlers from the DOM node.
 *
 * @param node DOM Node.
 * @returns Event Handlers.
 */
export function getEventHandlersFromDOMNode(
  node: Element,
): Array<EventHandler | null> | EventHandler | undefined | null {
  return node._ev;
}
