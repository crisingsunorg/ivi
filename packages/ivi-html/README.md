# VNode factories for HTML and SVG elements

## HTML

```ts
function a(className?: string): VNode<HTMLAnchorElementProps | null>;
function abbr(className?: string): VNode<HTMLElementProps | null>;
function acronym(className?: string): VNode<HTMLElementProps | null>;
function address(className?: string): VNode<HTMLElementProps | null>;
function applet(className?: string): VNode<HTMLAppletElementProps | null>;
function area(className?: string): VNode<HTMLAreaElementProps | null>;
function article(className?: string): VNode<HTMLElementProps | null>;
function aside(className?: string): VNode<HTMLElementProps | null>;
function audio(className?: string): VNode<HTMLAudioElementProps | null>;
function b(className?: string): VNode<HTMLElementProps | null>;
function base(className?: string): VNode<HTMLBaseElementProps | null>;
function basefont(className?: string): VNode<HTMLBaseFontElementProps | null>;
function bdo(className?: string): VNode<HTMLElementProps | null>;
function big(className?: string): VNode<HTMLElementProps | null>;
function blockquote(className?: string): VNode<HTMLQuoteElementProps | null>;
function body(className?: string): VNode<HTMLBodyElementProps | null>;
function br(className?: string): VNode<HTMLBRElementProps | null>;
function button(className?: string): VNode<HTMLButtonElementProps | null>;
function canvas(className?: string): VNode<HTMLCanvasElementProps | null>;
function caption(className?: string): VNode<HTMLTableCaptionElementProps | null>;
function center(className?: string): VNode<HTMLElementProps | null>;
function cite(className?: string): VNode<HTMLElementProps | null>;
function code(className?: string): VNode<HTMLElementProps | null>;
function col(className?: string): VNode<HTMLTableColElementProps | null>;
function colgroup(className?: string): VNode<HTMLTableColElementProps | null>;
function datalist(className?: string): VNode<HTMLDataListElementProps | null>;
function dd(className?: string): VNode<HTMLElementProps | null>;
function del(className?: string): VNode<HTMLModElementProps | null>;
function dfn(className?: string): VNode<HTMLElementProps | null>;
function dir(className?: string): VNode<HTMLDirectoryElementProps | null>;
function div(className?: string): VNode<HTMLDivElementProps | null>;
function dl(className?: string): VNode<HTMLDListElementProps | null>;
function dt(className?: string): VNode<HTMLElementProps | null>;
function em(className?: string): VNode<HTMLElementProps | null>;
function embed(className?: string): VNode<HTMLEmbedElementProps | null>;
function fieldset(className?: string): VNode<HTMLFieldSetElementProps | null>;
function figcaption(className?: string): VNode<HTMLElementProps | null>;
function figure(className?: string): VNode<HTMLElementProps | null>;
function font(className?: string): VNode<HTMLFontElementProps | null>;
function footer(className?: string): VNode<HTMLElementProps | null>;
function form(className?: string): VNode<HTMLFormElementProps | null>;
function frame(className?: string): VNode<HTMLFrameElementProps | null>;
function frameset(className?: string): VNode<HTMLFrameSetElementProps | null>;
function h1(className?: string): VNode<HTMLHeadingElementProps | null>;
function h2(className?: string): VNode<HTMLHeadingElementProps | null>;
function h3(className?: string): VNode<HTMLHeadingElementProps | null>;
function h4(className?: string): VNode<HTMLHeadingElementProps | null>;
function h5(className?: string): VNode<HTMLHeadingElementProps | null>;
function h6(className?: string): VNode<HTMLHeadingElementProps | null>;
function head(className?: string): VNode<HTMLHeadElementProps | null>;
function header(className?: string): VNode<HTMLElementProps | null>;
function hgroup(className?: string): VNode<HTMLElementProps | null>;
function hr(className?: string): VNode<HTMLHRElementProps | null>;
function html(className?: string): VNode<HTMLHtmlElementProps | null>;
function i(className?: string): VNode<HTMLElementProps | null>;
function iframe(className?: string): VNode<HTMLIFrameElementProps | null>;
function img(className?: string): VNode<HTMLImageElementProps | null>;
function ins(className?: string): VNode<HTMLModElementProps | null>;
function isindex(className?: string): VNode<HTMLUnknownElementProps | null>;
function kbd(className?: string): VNode<HTMLElementProps | null>;
function keygen(className?: string): VNode<HTMLElementProps | null>;
function label(className?: string): VNode<HTMLLabelElementProps | null>;
function legend(className?: string): VNode<HTMLLegendElementProps | null>;
function li(className?: string): VNode<HTMLLIElementProps | null>;
function link(className?: string): VNode<HTMLLinkElementProps | null>;
function listing(className?: string): VNode<HTMLPreElementProps | null>;
function map(className?: string): VNode<HTMLMapElementProps | null>;
function mark(className?: string): VNode<HTMLElementProps | null>;
function menu(className?: string): VNode<HTMLMenuElementProps | null>;
function meta(className?: string): VNode<HTMLMetaElementProps | null>;
function meter(className?: string): VNode<HTMLMeterElementProps | null>;
function nav(className?: string): VNode<HTMLElementProps | null>;
function nextid(className?: string): VNode<HTMLUnknownElementProps | null>;
function nobr(className?: string): VNode<HTMLElementProps | null>;
function noframes(className?: string): VNode<HTMLElementProps | null>;
function noscript(className?: string): VNode<HTMLElementProps | null>;
function object(className?: string): VNode<HTMLObjectElementProps | null>;
function ol(className?: string): VNode<HTMLOListElementProps | null>;
function optgroup(className?: string): VNode<HTMLOptGroupElementProps | null>;
function option(className?: string): VNode<HTMLOptionElementProps | null>;
function p(className?: string): VNode<HTMLParagraphElementProps | null>;
function param(className?: string): VNode<HTMLParamElementProps | null>;
function picture(className?: string): VNode<HTMLPictureElementProps | null>;
function plaintext(className?: string): VNode<HTMLElementProps | null>;
function pre(className?: string): VNode<HTMLPreElementProps | null>;
function progress(className?: string): VNode<HTMLProgressElementProps | null>;
function q(className?: string): VNode<HTMLQuoteElementProps | null>;
function rt(className?: string): VNode<HTMLElementProps | null>;
function ruby(className?: string): VNode<HTMLElementProps | null>;
function s(className?: string): VNode<HTMLElementProps | null>;
function samp(className?: string): VNode<HTMLElementProps | null>;
function script(className?: string): VNode<HTMLScriptElementProps | null>;
function section(className?: string): VNode<HTMLElementProps | null>;
function select(className?: string): VNode<HTMLSelectElementProps | null>;
function small(className?: string): VNode<HTMLElementProps | null>;
function source(className?: string): VNode<HTMLSourceElementProps | null>;
function span(className?: string): VNode<HTMLSpanElementProps | null>;
function strike(className?: string): VNode<HTMLElementProps | null>;
function strong(className?: string): VNode<HTMLElementProps | null>;
function style(className?: string): VNode<HTMLStyleElementProps | null>;
function sub(className?: string): VNode<HTMLElementProps | null>;
function sup(className?: string): VNode<HTMLElementProps | null>;
function table(className?: string): VNode<HTMLTableElementProps | null>;
function tbody(className?: string): VNode<HTMLTableSectionElementProps | null>;
function td(className?: string): VNode<HTMLTableDataCellElementProps | null>;
function template(className?: string): VNode<HTMLTemplateElementProps | null>;
function textarea(className?: string): VNode<HTMLTextAreaElementProps | null>;
function tfoot(className?: string): VNode<HTMLTableSectionElementProps | null>;
function th(className?: string): VNode<HTMLTableHeaderCellElementProps | null>;
function thead(className?: string): VNode<HTMLTableSectionElementProps | null>;
function title(className?: string): VNode<HTMLTitleElementProps | null>;
function tr(className?: string): VNode<HTMLTableRowElementProps | null>;
function track(className?: string): VNode<HTMLTrackElementProps | null>;
function tt(className?: string): VNode<HTMLElementProps | null>;
function u(className?: string): VNode<HTMLElementProps | null>;
function ul(className?: string): VNode<HTMLUListElementProps | null>;
function video(className?: string): VNode<HTMLVideoElementProps | null>;
function wbr(className?: string): VNode<HTMLElementProps | null>;
function x_ms_webview(className?: string): VNode<MSHTMLWebViewElementProps | null>;
function xmp(className?: string): VNode<HTMLPreElementProps | null>;
```

### Input Elements

```ts
function inputButton(className?: string): VNode<HTMLInputElementProps | null>;
function inputCheckbox(className?: string): VNode<HTMLInputElementProps | null>;
function inputColor(className?: string): VNode<HTMLInputElementProps | null>;
function inputDate(className?: string): VNode<HTMLInputElementProps | null>;
function inputDatetime(className?: string): VNode<HTMLInputElementProps | null>;
function inputDatetimeLocal(className?: string): VNode<HTMLInputElementProps | null>;
function inputEmail(className?: string): VNode<HTMLInputElementProps | null>;
function inputFile(className?: string): VNode<HTMLInputElementProps | null>;
function inputHidden(className?: string): VNode<HTMLInputElementProps | null>;
function inputImage(className?: string): VNode<HTMLInputElementProps | null>;
function inputMonth(className?: string): VNode<HTMLInputElementProps | null>;
function inputNumber(className?: string): VNode<HTMLInputElementProps | null>;
function inputPassword(className?: string): VNode<HTMLInputElementProps | null>;
function inputRadio(className?: string): VNode<HTMLInputElementProps | null>;
function inputRange(className?: string): VNode<HTMLInputElementProps | null>;
function inputReset(className?: string): VNode<HTMLInputElementProps | null>;
function inputSearch(className?: string): VNode<HTMLInputElementProps | null>;
function inputSubmit(className?: string): VNode<HTMLInputElementProps | null>;
function inputTel(className?: string): VNode<HTMLInputElementProps | null>;
function inputText(className?: string): VNode<HTMLInputElementProps | null>;
function inputTime(className?: string): VNode<HTMLInputElementProps | null>;
function inputUrl(className?: string): VNode<HTMLInputElementProps | null>;
function inputWeek(className?: string): VNode<HTMLInputElementProps | null>;
```

## SVG

```ts
function circle(className?: string): VNode<SVGCircleElementProps | null>;
function clippath(className?: string): VNode<SVGClipPathElementProps | null>;
function defs(className?: string): VNode<SVGDefsElementProps | null>;
function desc(className?: string): VNode<SVGDescElementProps | null>;
function ellipse(className?: string): VNode<SVGEllipseElementProps | null>;
function feblend(className?: string): VNode<SVGFEBlendElementProps | null>
function fecolormatrix(className?: string): VNode<SVGFEColorMatrixElementProps | null>;
function fecomponenttransfer(className?: string): VNode<SVGFEComponentTransferElementProps | null>;
function fecomposite(className?: string): VNode<SVGFECompositeElementProps | null>;
function feconvolvematrix(className?: string): VNode<SVGFEConvolveMatrixElementProps | null>;
function fediffuselighting(className?: string): VNode<SVGFEDiffuseLightingElementProps | null>;
function fedisplacementmap(className?: string): VNode<SVGFEDisplacementMapElementProps | null>;
function fedistantlight(className?: string): VNode<SVGFEDistantLightElementProps | null>;
function feflood(className?: string): VNode<SVGFEFloodElementProps | null>;
function fefunca(className?: string): VNode<SVGFEFuncAElementProps | null>;
function fefuncb(className?: string): VNode<SVGFEFuncBElementProps | null>;
function fefuncg(className?: string): VNode<SVGFEFuncGElementProps | null>;
function fefuncr(className?: string): VNode<SVGFEFuncRElementProps | null>;
function fegaussianblur(className?: string): VNode<SVGFEGaussianBlurElementProps | null>;
function feimage(className?: string): VNode<SVGFEImageElementProps | null>;
function femerge(className?: string): VNode<SVGFEMergeElementProps | null>;
function femergenode(className?: string): VNode<SVGFEMergeNodeElementProps | null>;
function femorphology(className?: string): VNode<SVGFEMorphologyElementProps | null>;
function feoffset(className?: string): VNode<SVGFEOffsetElementProps | null>;
function fepointlight(className?: string): VNode<SVGFEPointLightElementProps | null>;
function fespecularlighting(className?: string): VNode<SVGFESpecularLightingElementProps | null>;
function fespotlight(className?: string): VNode<SVGFESpotLightElementProps | null>;
function fetile(className?: string): VNode<SVGFETileElementProps | null>;
function feturbulence(className?: string): VNode<SVGFETurbulenceElementProps | null>;
function filter(className?: string): VNode<SVGFilterElementProps | null>;
function foreignobject(className?: string): VNode<SVGForeignObjectElementProps | null>;
function g(className?: string): VNode<SVGGElementProps | null>;
function image(className?: string): VNode<SVGImageElementProps | null>;
function line(className?: string): VNode<SVGLineElementProps | null>;
function lineargradient(className?: string): VNode<SVGLinearGradientElementProps | null>;
function marker(className?: string): VNode<SVGMarkerElementProps | null>;
function mask(className?: string): VNode<SVGMaskElementProps | null>;
function metadata(className?: string): VNode<SVGMetadataElementProps | null>;
function path(className?: string): VNode<SVGPathElementProps | null>;
function pattern(className?: string): VNode<SVGPatternElementProps | null>;
function polygon(className?: string): VNode<SVGPolygonElementProps | null>;
function polyline(className?: string): VNode<SVGPolylineElementProps | null>;
function radialgradient(className?: string): VNode<SVGRadialGradientElementProps | null>;
function rect(className?: string): VNode<SVGRectElementProps | null>;
function stop(className?: string): VNode<SVGStopElementProps | null>;
function svg(className?: string): VNode<SVGSVGElementProps | null>;
function symbol(className?: string): VNode<SVGSymbolElementProps | null>;
function text(className?: string): VNode<SVGTextElementProps | null>;
function textpath(className?: string): VNode<SVGTextPathElementProps | null>;
function tspan(className?: string): VNode<SVGTSpanElementProps | null>;
function use(className?: string): VNode<SVGUseElementProps | null>;
function view(className?: string): VNode<SVGViewElementProps | null>;
```