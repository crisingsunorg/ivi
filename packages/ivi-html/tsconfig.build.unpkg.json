{
  "extends": "./tsconfig.build.module.json",
  "compilerOptions": {
    "preserveConstEnums": true
  }
}
