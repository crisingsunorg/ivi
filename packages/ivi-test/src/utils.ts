import { EventHandler, EventSource } from "ivi-events";
import { VNode, VNodeFlags } from "ivi";

/**
 * containsClassName checks if className list contains a className.
 *
 * @param classNames className list in a HTML format with all values separated by space characters.
 * @param className className to check.
 * @returns true when className list contains a className.
 */
export function containsClassName(classNames: string, className: string): boolean {
  return classNames
    .split(" ")
    .map((c) => c.trim())
    .some((c) => c === className);
}

/**
 * matchValues checks property values for matching. Properties that aren't in `match` property are ignored.
 *
 * @param props Properties that should be matched against `match` property.
 * @param match Is a map of values that should match.
 * @returns true when properties are matching.
 */
export function matchValues(
  props: { [key: string]: any } | null,
  match: { [key: string]: any } | null,
): boolean {
  if (match !== null) {
    const keys = Object.keys(match);
    if (keys.length > 0) {
      if (props === null) {
        return false;
      }
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        if (props.hasOwnProperty(key) === false || props[key] !== match[key]) {
          return false;
        }
      }
    }
  }
  return true;
}

/**
 * matchKeys checks that property keys are existing.
 *
 * @param props Properties that should be checked against `match` property.
 * @param match Is a map of keys that should exist.
 * @returns true when matching properties are existing.
 */
export function matchKeys(
  props: { [key: string]: any } | null,
  match: { [key: string]: boolean },
): boolean {
  const keys = Object.keys(match);
  if (keys.length > 0) {
    if (props === null) {
      return false;
    }
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i];
      if (match[key] !== props.hasOwnProperty(key)) {
        return false;
      }
    }
  }
  return true;
}

/**
 * containsEventHandler checks if event handler list contain an event source.
 *
 * @param eventHandlers Event handlers.
 * @param eventSource Event source.
 * @returns true when event handlers contain an event source.
 */
export function containsEventHandler(
  eventHandlers: Array<EventHandler | null> | EventHandler | null,
  eventSource: EventSource,
): boolean {
  if (eventHandlers !== null) {
    if (Array.isArray(eventHandlers)) {
      for (let i = 0; i < eventHandlers.length; i++) {
        const h = eventHandlers[i];
        if (h !== null && h.source === eventSource) {
          return true;
        }
      }
    } else {
      return (eventHandlers.source === eventSource);
    }
  }
  return false;
}

const VNodeLooseMatchFlags = 0
  | VNodeFlags.Text
  | VNodeFlags.Element
  | VNodeFlags.ComponentFunction
  | VNodeFlags.ComponentClass
  | VNodeFlags.InputElement
  | VNodeFlags.TextAreaElement
  | VNodeFlags.MediaElement
  | VNodeFlags.SvgElement
  | VNodeFlags.Connect
  | VNodeFlags.UpdateContext
  | VNodeFlags.KeepAlive
  | VNodeFlags.VoidElement;

/**
 * isVNodeLooseMatch performs a loose match on VNodes.
 *
 * @param a VNode.
 * @param b VNode.
 * @returns true when VNode are loosely matching.
 */
export function isVNodeLooseMatch(a: VNode<any>, b: VNode<any>): boolean {
  const bFlags = b._flags;
  if (((a._flags ^ bFlags) & VNodeLooseMatchFlags) !== 0) {
    return false;
  }

  if (a._tag !== b._tag) {
    return false;
  }

  if ((bFlags & VNodeFlags.Element) !== 0) {
    if (b._props !== null) {
      if (a._props === null) {
        return false;
      }

      if (matchValues(a._props.attrs, b._props.attrs) === false) {
        return false;
      }
      if (matchValues(a._props.syle, b._props.style) === false) {
        return false;
      }

    }
  }
  return true;
}
