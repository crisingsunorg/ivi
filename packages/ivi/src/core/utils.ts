/**
 * mut mutates an object.
 *
 * @example
 *
 *     interface Entry {
 *       _dirty: number;
 *       title: string;
 *     }
 *     const m = (entry: Entry) => { entry._dirty = dirty(); };
 *     const useEntry = selector((entry: Entry) => entry._dirty);
 *     const entrySetTitle = mut(m, (entry: Entry, title: string) => { entry.title = title; });
 *
 *     const EntryView = component<Entry>((c) => {
 *       const dirtyCheckEntry = useEntry(c);
 *
 *       return (entry) => (
 *         dirtyCheckEntry(entry),
 *
 *         div().t(entry.title)
 *       )
 *     });
 *
 * @param s - Selector function.
 * @param shouldUpdate - Should update function.
 * @returns selector factory.
 */
export const mut = <T, U extends any[], V>(d: (v: T, flags?: V) => void, fn: (v: T, ...args: U) => void | V) => (
  function () {
    const r = fn.apply(void 0, arguments);
    if (r === void 0 || r > 0) {
      d(arguments[0], r);
    }
  } as (v: T, ...args: U) => void | V
);
