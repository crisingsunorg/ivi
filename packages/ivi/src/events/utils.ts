import { PASSIVE_EVENTS } from "ivi-core";

/* istanbul ignore next */
/**
 * getEventTarget retrieves target element from an event.
 *
 * There are still some differences in modern browsers, so we need to use this function to retrieve event targets.
 *
 * #quirks
 *
 * @param ev Event.
 * @returns Target Element.
 */
export function getEventTarget(ev: Event): EventTarget {
  let target = ev.target || window;

  /**
   * Some browsers are implementing it according to SVG 1.1 specs:
   *
   * https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7998724/
   *
   * "If event attributes are assigned to referenced elements, then the actual target for the event will be the
   * SVGElementInstance object within the 'instance tree' corresponding to the given referenced element."
   *
   * https://www.w3.org/TR/SVG/struct.html#UseElement
   *
   * SVG 2 redefined the use event handling model:
   *
   * - Removed the SVGElementInstance and SVGElementInstanceList interfaces, and the corresponding attributes on the
   *   SVGUseElement interface.
   * - Changed the ‘use’ element event flow to follow the Shadow DOM spec.
   *
   * https://www.w3.org/TR/SVG2/changes.html#structure
   */
  if ((target as any).correspondingUseElement !== undefined) {
    target = (target as any).correspondingUseElement;
  }

  return target;
}

/* istanbul ignore next */
/**
 * `{ capture: true, passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": true } : true;

/* istanbul ignore next */
/**
 * `{ capture: true, passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": false } : true;

/* istanbul ignore next */
/**
 * `{ passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": true } : false;

/* istanbul ignore next */
/**
 * `{ passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": false } : false;
