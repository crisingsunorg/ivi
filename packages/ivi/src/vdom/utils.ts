import { Box } from "ivi-shared";
import { OpState } from "./state";
import { Component } from "./component";
import { useSelect } from "./hooks";
import { getDOMNode } from "./reconciler";

/**
 * selector creates a selector factory.
 *
 * @example
 *
 *     interface Entry {
 *       title: string;
 *     }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @returns Selector factory.
 */
export function selector<T>(
  s: () => T,
): (component: Component) => () => T;

/**
 * selector creates a selector factory.
 *
 * @example
 *
 * interface Entry {
 *   title: string;
 * }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @param shouldUpdate Should update function.
 * @returns Selector factory.
 */
export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T;

export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T {
  return (component: Component) => useSelect(component, s, shouldUpdate);
}

/**
 * findDOMNode finds a closest DOM node.
 *
 * @typeparam T DOM Node type.
 * @param box Boxed op node state.
 * @returns DOM Node or a `null` value.
 */
export const findDOMNode = <T extends Node>(
  box: Box<OpState | null>,
) => box.v === null ? null : getDOMNode(box.v) as T;
