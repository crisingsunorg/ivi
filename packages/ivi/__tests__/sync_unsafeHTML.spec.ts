import * as h from "ivi-html";
import { startRender, checkDOMOps, domOps } from "./utils";

test(`<div> => <div unsafeHTML="<span>abc</span>"></div>`, () => {
  startRender((r) => {
    checkDOMOps((c) => {
      r(h.div());
      const n = r(h.div().unsafeHTML("<span>abc</span>"));

      expect(n).toMatchSnapshot();
      expect(c).toEqual(domOps(2, 0, 0, 0, 1, 1, 0));
    });
  });
});

test(`<div unsafeHTML="<div>abc</div>"> => <div></div>`, () => {
  startRender((r) => {
    checkDOMOps((c) => {
      r(h.div().unsafeHTML("<div>abc</div>"));
      const n = r(h.div());

      expect(n).toMatchSnapshot();
      expect(c).toEqual(domOps(2, 0, 0, 0, 1, 1, 0));
    });
  });
});

test(`<div unsafeHTML={ null }> => <div unsafeHTML="<span>abc</span>"></div>`, () => {
  startRender((r) => {
    checkDOMOps((c) => {
      r(h.div().unsafeHTML(null));
      const n = r(h.div().unsafeHTML("<span>abc</span>"));

      expect(n).toMatchSnapshot();
      expect(c).toEqual(domOps(1, 0, 0, 0, 1, 0, 0));
    });
  });
});

test(`<div unsafeHTML="<div>abc</div>"> => <div unsafeHTML={ null }></div>`, () => {
  startRender((r) => {
    checkDOMOps((c) => {
      r(h.div().unsafeHTML("<div>abc</div>"));
      const n = r(h.div().unsafeHTML(null));

      expect(n).toMatchSnapshot();
      expect(c).toEqual(domOps(1, 0, 0, 0, 1, 0, 0));
    });
  });
});

test(`<div unsafeHTML="<div>abc</div>"> => <div unsafeHTML="<span>abc</span>"></div>`, () => {
  startRender((r) => {
    checkDOMOps((c) => {
      r(h.div().unsafeHTML("<div>abc</div>"));
      const n = r(h.div().unsafeHTML("<span>abc</span>"));

      expect(n).toMatchSnapshot();
      expect(c).toEqual(domOps(1, 0, 0, 0, 1, 0, 0));
    });
  });
});
