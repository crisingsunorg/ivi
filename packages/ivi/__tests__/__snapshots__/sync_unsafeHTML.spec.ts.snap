// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<div unsafeHTML="<div>abc</div>"> => <div unsafeHTML="<span>abc</span>"></div> 1`] = `
<div>
  <span>
    abc
  </span>
</div>
`;

exports[`<div unsafeHTML="<div>abc</div>"> => <div unsafeHTML={ null }></div> 1`] = `<div />`;

exports[`<div unsafeHTML="<div>abc</div>"> => <div></div> 1`] = `<div />`;

exports[`<div unsafeHTML={ null }> => <div unsafeHTML="<span>abc</span>"></div> 1`] = `
<div>
  <span>
    abc
  </span>
</div>
`;

exports[`<div> => <div unsafeHTML="<span>abc</span>"></div> 1`] = `
<div>
  <span>
    abc
  </span>
</div>
`;
